src/distance/binary_quantized_euclidean.rs

```diff
@@ -0,0 +1,79 @@
+use std::borrow::Cow;
+
+use bytemuck::{Pod, Zeroable};
+use rand::Rng;
+
+use super::two_means;
+use crate::distance::Distance;
+use crate::node::Leaf;
+use crate::parallel::ImmutableSubsetLeafs;
+use crate::unaligned_vector::{self, BinaryQuantized, UnalignedVector};
+
+/// The Euclidean distance between two points in Euclidean space
+/// is the length of the line segment between them.
+///
+/// `d(p, q) = sqrt((p - q)²)`
+#[derive(Debug, Clone)]
+pub enum BinaryQuantizedEuclidean {}
+
+/// The header of BinaryQuantizedEuclidean leaf nodes.
+#[repr(C)]
+#[derive(Pod, Zeroable, Debug, Clone, Copy)]
+pub struct NodeHeaderBinaryQuantizedEuclidean {
+    /// An extra constant term to determine the offset of the plane
+    bias: f32,
+}
+
+impl Distance for BinaryQuantizedEuclidean {
+    type Header = NodeHeaderBinaryQuantizedEuclidean;
+    type VectorFormat = unaligned_vector::BinaryQuantized;
+
+    fn name() -> &'static str {
+        "binary quantized euclidean"
+    }
+
+    fn new_header(_vector: &UnalignedVector<Self::VectorFormat>) -> Self::Header {
+        NodeHeaderBinaryQuantizedEuclidean { bias: 0.0 }
+    }
+
+    fn built_distance(p: &Leaf<Self>, q: &Leaf<Self>) -> f32 {
+        dot_product(&p.vector, &q.vector)
+    }
+
+    fn norm_no_header(v: &UnalignedVector<Self::VectorFormat>) -> f32 {
+        dot_product(v, v).sqrt()
+    }
+
+    fn init(_node: &mut Leaf<Self>) {}
+
+    fn create_split<'a, R: Rng>(
+        children: &'a ImmutableSubsetLeafs<Self>,
+        rng: &mut R,
+    ) -> heed::Result<Cow<'a, UnalignedVector<Self::VectorFormat>>> {
+        let [node_p, node_q] = two_means(rng, children, false)?;
+        let vector: Vec<f32> =
+            node_p.vector.iter().zip(node_q.vector.iter()).map(|(p, q)| p - q).collect();
+        let mut normal = Leaf {
+            header: NodeHeaderBinaryQuantizedEuclidean { bias: 0.0 },
+            vector: UnalignedVector::from_slice(&vector),
+        };
+        Self::normalize(&mut normal);
+
+        Ok(Cow::Owned(normal.vector.into_owned()))
+    }
+
+    fn margin(p: &Leaf<Self>, q: &Leaf<Self>) -> f32 {
+        p.header.bias + dot_product(&p.vector, &q.vector)
+    }
+
+    fn margin_no_header(
+        p: &UnalignedVector<Self::VectorFormat>,
+        q: &UnalignedVector<Self::VectorFormat>,
+    ) -> f32 {
+        dot_product(p, q)
+    }
+}
+
+fn dot_product(u: &UnalignedVector<BinaryQuantized>, v: &UnalignedVector<BinaryQuantized>) -> f32 {
+    u.as_bytes().iter().zip(v.as_bytes()).map(|(u, v)| (u ^ v).count_ones()).sum::<u32>() as f32
```

Comment:
Could you put a comment that relevancy can be impacted here? But for non-module 64 dimensions.